### PR TITLE
[VPlan] Model FOR resume value extraction in VPlan.

### PR DESCRIPTION
**Pull request by fhahn**
This patch uses the ExtractFromEnd VPInstruction opcode
to extract the value of a FOR to be used as resume value for the ph in
the scalar loop.

It adds a new live-out that temporarily wraps the FOR phi in the scalar loop.
fixFixedOrderRecurrence will process live outs for fixed order recurrence phis
by creating a new phi node in the scalar preheader, using the generated value
for the live-out as incoming value from the middle block and the original start
value as incoming value for the other edge.

Depends on https://github.com/llvm/llvm-project/pull/93395

### CODE REVIEW

**Comment by fhahn** [comment]
Updated based on the latest version of https://github.com/llvm/llvm-project/pull/93395, this now simply uses the `ExtractFromEnd` opcode. 

This leaves handling the live-out that's used in the scalar pre-header. To do so, it adds live-outs for the scalar loops header phis, which later get fixed up with the newly created phis for the resume values

**Review by fhahn — COMMENTED** [review]
Updated now that https://github.com/llvm/llvm-project/pull/93395 landed

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 1]
```diff
@@ -855,6 +904,13 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
         {}, "vector.recur.extract.for.phi"));
     RecurSplice->replaceUsesWithIf(
         Result, [](VPUser &U, unsigned) { return isa<VPLiveOut>(&U); });
+    auto *Resume = MiddleBuilder.createNaryOp(
+        VPInstruction::ExtractFromEnd,
+        {FOR->getBackedgeValue(),
+         Plan.getOrAddLiveIn(ConstantInt::get(IntTy, 1))},
+        {}, "vector.recur.extract");
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:911]: These two names `vector.recur.extract.for.phi` and `vector.recur.extract` should also be improved - both are extracting from the same vector - the former extracts the final value of the recurrence for uses outside the loop while the latter extract a value for the next iteration to continue the recurrence.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -3391,16 +3390,16 @@ void InnerLoopVectorizer::fixVectorizedLoop(VPTransformState &State,
...
-  VPBasicBlock *HeaderVPBB = VectorRegion->getEntryBasicBlock();
-  for (VPRecipeBase &R : HeaderVPBB->phis()) {
-    if (auto *FOR = dyn_cast<VPFirstOrderRecurrencePHIRecipe>(&R))
-      fixFixedOrderRecurrence(FOR, State);
+  // vector form. Note that fixing reduction phis, as well as extracting the
+  // exit and resume values for fixed-order recurrences are already modeled in
+  // VPlan. All that remains to do here is creating a phi in the scalar
+  // pre-header for each fixed-rder recurrence resume value.
+  // TODO: Also model creating phis in the scalar pre-header in VPlan.
+  for (const auto &[_, LO] : to_vector(Plan.getLiveOuts())) {
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3398]: Removing LiveOuts while iterating over them requires the `to_vector`? Suffice to make_early_inc_range?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -3391,16 +3390,16 @@ void InnerLoopVectorizer::fixVectorizedLoop(VPTransformState &State,
...
+    if (!Legal->isFixedOrderRecurrence(LO->getPhi()))
+      continue;
+    fixFixedOrderRecurrence(LO, State);
+    Plan.removeLiveOut(LO->getPhi());
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3402]: Worth a comment, somewhere. Plan holds LO's that need fixing, any LO that is fixed is removed from Plan, remaining LO's receive a default LO->fixPhi() fix below.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 4]
```diff
@@ -847,6 +847,55 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
...
+    //     x = v2(3)
+    //     br scalar.ph
+    //
+    //   scalar.ph:
+    //     s_init = phi [x, middle.block], [a[-1], otherwise]
+    //     br scalar.body
+    //
+    // After execution completes the vector loop, we extract the next value of
+    // the recurrence (x) to use as the initial value in the scalar loop. This
+    // is modeled by ExtractRecurrenceResume.
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:898]:
```suggestion
    // is modeled by ExtractFromEnd.
```

**Review by ayalz — COMMENTED** [review]
Very nice! Raises some thoughts, potentially for subsequent or preliminary patches.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 5]
```diff
@@ -855,6 +904,13 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
         {}, "vector.recur.extract.for.phi"));
     RecurSplice->replaceUsesWithIf(
         Result, [](VPUser &U, unsigned) { return isa<VPLiveOut>(&U); });
+    auto *Resume = MiddleBuilder.createNaryOp(
+        VPInstruction::ExtractFromEnd,
+        {FOR->getBackedgeValue(),
+         Plan.getOrAddLiveIn(ConstantInt::get(IntTy, 1))},
+        {}, "vector.recur.extract");
+    // Introduce VPUsers modeling the exit values.
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Updated both the comment + commit message, thanks! Not sure how to better keep the distinction, but I added comment where we generate the extract for the exit.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 6]
```diff
@@ -3469,85 +3470,28 @@ void InnerLoopVectorizer::fixVectorizedLoop(VPTransformState &State,
...
 
   // Fix the initial value of the original recurrence in the scalar loop.
   Builder.SetInsertPoint(LoopScalarPreHeader, LoopScalarPreHeader->begin());
-  PHINode *Phi = cast<PHINode>(PhiR->getUnderlyingValue());
-  auto *Start = Builder.CreatePHI(Phi->getType(), 2, "scalar.recur.init");
-  auto *ScalarInit = PhiR->getStartValue()->getLiveInIRValue();
+  PHINode *ScalarHeaderPhi = LO->getPhi();
+  auto *NewScalarHeaderPhi = Builder.CreatePHI(ScalarHeaderPhi->getType(), 2, "scalar.recur.init");
+  auto *InitScalarFOR =
+      ScalarHeaderPhi->getIncomingValueForBlock(LoopScalarPreHeader);
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:3487]:
```suggestion
  PHINode *ScalarHeaderPhi = LO->getPhi();
  auto *InitScalarFOR =
      ScalarHeaderPhi->getIncomingValueForBlock(LoopScalarPreHeader);

  Builder.SetInsertPoint(LoopScalarPreHeader, LoopScalarPreHeader->begin());
  auto *NewScalarHeaderPhi = Builder.CreatePHI(ScalarHeaderPhi->getType(), 2, "scalar.recur.init");
```
order slightly more consistent?

`NewScalarHeaderPhi` is created in the preheader, better named `ScalarPreheaderPhi`?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 7]
```diff
@@ -3166,7 +3166,7 @@ class VPlan {
   /// definitions are VPValues that hold a pointer to their underlying IR.
   SmallVector<VPValue *, 16> VPLiveInsToFree;
 
-  /// Values used outside the plan.
+  /// Values used outside the plan. It contains live-outs that need fixing. Any live-out that is fixed outside VPlan needs to be removed. The remaining live-outs are fixed via VPLiveOut::fixPhi.
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:3169]: clang-format?

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 8]
```diff
@@ -847,14 +847,84 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
     // all users.
     RecurSplice->setOperand(0, FOR);
 
+    // This is the second phase of vectorizing first-order recurrences. An
+    // overview of the transformation is described below. Suppose we have the
+    // following loop.
+    //
+    //   for (int i = 0; i < n; ++i)
+    //     b[i] = a[i] - a[i - 1];
+    //
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:856]:
```suggestion
    //
    // with some use after the loop of the last a[i-1], e.g.,
    //
    //   for (int i = 0; i < n; ++i) {
    //     t = a[i - 1];
    //     b[i] = a[i] - t;
    //   }
    //    use t;
    //
```

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 9]
```diff
@@ -847,14 +847,84 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
...
+    //     s_init = a[-1]
+    //     br scalar.body
+    //
+    //   scalar.body:
+    //     i = phi [0, scalar.ph], [i+1, scalar.body]
+    //     s1 = phi [s_init, scalar.ph], [s2, scalar.body]
+    //     s2 = a[i]
+    //     b[i] = s2 - s1
+    //     br cond, scalar.body, ...
+    //
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:870]:
```suggestion
    //     br cond, scalar.body, exit.block
    //
    //   exit.block:
    //     use = lcssa.phi [s1, scalar.body]
    //
```

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 10]
```diff
@@ -847,14 +847,84 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
...
+    //     i = phi [0, vector.ph], [i+4, vector.body]
+    //     v1 = phi [v_init, vector.ph], [v2, vector.body]
+    //     v2 = a[i, i+1, i+2, i+3];
+    //     v3 = vector(v1(3), v2(0, 1, 2))
+    //     b[i, i+1, i+2, i+3] = v2 - v3
+    //     br cond, vector.body, middle.block
+    //
+    //   middle.block:
+    //     s_penultimate = v2(2) = v3(3)
+		//     s_resume = v2(3)
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:890]:
```suggestion
    //     s_penultimate = v2(2) = v3(3)
    //     s_resume = v2(3)
```

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 1 above)
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:926]: Better retain names in this PR, and change them in a separate patch either before or after.

**Review by ayalz — APPROVED** [review]
This looks good to me, thanks! Added last minor comments, including a need for clang-format. Commit message can add that the phi in the scalar pre-header that routes the resume value is destined to be modeled in VPlan,

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 6 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Adjusted, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 7 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Yep, applied clang-format and fixed various formatting issue, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 8 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:859]: Merged with the example loop, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 9 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Added, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 10 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Fixed, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:933]: 👍🏻

**Pull request merged**